crates/sui-indexer-alt/src/pipeline/concurrent/collector.rs

```diff
@@ -179,3 +182,166 @@ pub(super) fn collector<H: Handler + 'static>(
+    impl Processor for TestHandler {
+        type Value = u64;
+        const NAME: &'static str = "test_handler";
+        const FANOUT: usize = 1;
+
+        fn process(&self, _checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
+            Ok(vec![])
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl Handler for TestHandler {
+        const MIN_EAGER_ROWS: usize = 3;
+        const MAX_CHUNK_ROWS: usize = 4;
+        const MAX_PENDING_ROWS: usize = 10;
+
+        async fn commit(
+            _values: &[Self::Value],
+            _conn: &mut db::Connection<'_>,
+        ) -> anyhow::Result<usize> {
+            tokio::time::sleep(Duration::from_millis(1000)).await;
+            Ok(0)
+        }
+    }
+
+    #[tokio::test]
+    async fn test_collector_batches_data() {
+        let (processor_tx, processor_rx) = mpsc::channel(10);
+        let (collector_tx, mut collector_rx) = mpsc::channel(10);
+        let metrics = Arc::new(IndexerMetrics::new(&Registry::new()));
+        let cancel = CancellationToken::new();
+
+        let _collector = collector::<TestHandler>(
+            PipelineConfig::default(),
+            processor_rx,
+            collector_tx,
+            metrics,
+            cancel.clone(),
+        );
+
+        // Send test data
+        let test_data = vec![
+            Indexed::new(0, 1, 10, 1000, vec![1, 2]),
+            Indexed::new(0, 2, 20, 2000, vec![3, 4]),
+            Indexed::new(0, 3, 30, 3000, vec![5, 6]),
+        ];
+
+        for data in test_data {
+            processor_tx.send(data).await.unwrap();
+        }
+
+        let batch1 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch1.len(), 4);
+
+        let batch2 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch2.len(), 2);
+
+        let batch3 = collector_rx.recv().await.unwrap();
+        assert_eq!(batch3.len(), 0);
+
+        cancel.cancel();
+    }
+
+    #[tokio::test]
+    async fn test_collector_shutdown() {
+        let (processor_tx, processor_rx) = mpsc::channel(10);
+        let (collector_tx, mut collector_rx) = mpsc::channel(10);
+        let metrics = Arc::new(IndexerMetrics::new(&Registry::new()));
+        let cancel = CancellationToken::new();
+
+        let _collector = collector::<TestHandler>(
+            PipelineConfig::default(),
+            processor_rx,
+            collector_tx,
+            metrics,
+            cancel.clone(),
+        );
+
+        processor_tx
+            .send(Indexed::new(0, 1, 10, 1000, vec![1, 2]))
+            .await
+            .unwrap();
+
+        let batch = collector_rx.recv().await.unwrap();
+        assert_eq!(batch.len(), 2);
+
+        // Drop processor sender to simulate shutdown
+        drop(processor_tx);
+
+        // After a short delay, collector should shut down
+        tokio::time::sleep(Duration::from_millis(200)).await;
+        assert!(collector_rx.try_recv().is_err());
+
+        cancel.cancel();
```

Comment:
Rather than having an explicit delay, you can use a timeout here. That way, if the collector shuts down sooner, your test will finish sooner, but you will also not hang the test if it fails (this also means you can afford to use a longer timeout to make the test less flaky without worrying about it increasing runtimes all the time):

```suggestion
        // After a short delay, collector should shut down
        tokio::time::timeout(Duration::from_millis(500), collector)
          .await
          .expect("collector did not shutdown");
          
        assert!(collector_rx.try_recv().is_err());
```